[Clarissa Lee Zi Xin] iP

clarlzx opened the pull request:
**What is the purpose of this pull request?**

- Chat the Cat replaces Duke as the personality behind the chat bot
- Additional functionalities are added to Chat the Cat

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

clarlzx committed:
Level-1: Greet, Echo, Exit

clarlzx committed:
Level-2: Add, List

clarlzx committed:
Level-3: Mark as Done, Task class added

clarlzx committed:
Level-4: Todo, Event, Deadline, added Todo/Deadline/Event class

clarlzx committed:
A-TextUiTesting: Automated Text UI Testing

clarlzx committed:
Level-5: Handle Errors, added ChatException class

clarlzx committed:
Level-6: Delete

clarlzx committed:
Level-7: Save and load tasks from hard disk

clarlzx committed:
Level-8: Store dates and times as LocalDateTime object

clarlzx committed:
Level-8: Add more test cases

clarlzx committed:
Merge branch 'branch-Level-7' into master

clarlzx committed:
Merge branch 'branch-Level-8' into master

clarlzx committed:
Level-8: Fix errors

clarlzx committed:
A-MoreOOP: Add Command classes, Parser, Storage, Ui and TaskList classes

clarlzx committed:
Rename Duke.java to Chat.java

clarlzx committed:
Rename Duke in files to Chat

clarlzx committed:
A-Packages: Sort all classes into suitable packages

clarlzx committed:
Merge branch 'add-gradle-support' of https://github.com/clarlzx/ip into master

clarlzx committed:
A-JUnit: Add JUnit tests for Todo, Event and Deadline

clarlzx committed:
A-JUnit: Edit names of some tests

clarlzx committed:
A-CodingStandard: Add Javadocs comments for all classes

clarlzx committed:
A-CodingStandard: Edit code to follow coding standard

clarlzx committed:
Level-9: Add FindCommand class and methods

clarlzx committed:
Merge branch 'branch-A-JavaDoc' into master

clarlzx committed:
Merge branch 'branch-A-CodingStandard' into master

clarlzx committed:
Merge branch 'branch-Level-9' into master

clarlzx committed:
Level-9: Fix conflict errors for FindCommand class

Level-9: Add JavaDoc comments to FindCommand class

JodyLorah commented on src/main/java/chat/Parser.java:
```diff
@@ -1,11 +1,6 @@
-import chat.command.DeleteCommand;
-import chat.command.AddCommand;
-import chat.command.ExitCommand;
+import chat.command.*;
```
Maybe list out all imports explicitly instead of *

clarlzx replied:
Yes that seems better, thank you for pointing that out.

JodyLorah commented on src/main/java/chat/command/FindCommand.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.ArrayList;
+
+public class FindCommand extends Command {
```
Maybe add some javadocs so others would know what youre trying to do in this class

clarlzx replied:
Oh didn't realize that I missed that out, thank you for pointing it out.

JodyLorah commented on src/main/java/chat/Storage.java:
```diff
@@ -56,7 +56,7 @@ public void save(TaskList tasks) throws ChatException {
     public ArrayList<Task> load() throws ChatException {
         try {
-            ArrayList<Task> taskListFromFile = new ArrayList<>();
+            ArrayList<Task> tasksFromFile = new ArrayList<>();
```
Nice use of plural since it is an arraylist of tasks

JodyLorah commented on src/main/java/chat/task/Task.java:
```diff
@@ -30,20 +30,20 @@ public String getName() {
      */
-    public boolean getDone() {
-        return this.done;
+    public boolean getIsDone() {
```
Maybe name the boolean as a question isTaskDone

JodyLorah commented on src/main/java/chat/task/Todo.java:
```diff
@@ -62,13 +62,13 @@ public static Todo createTodo (String str) throws ChatException {
      * Returns a string that shows the details of the task.
-     * <p>[ ] will be displayed if done = false.</p>
-     * <p>[X] will be displayed if done = true.</p>
+     * <p>[ ] will be displayed if isDone = false.</p>
+     * <p>[X] will be displayed if isDone = true.</p>
```
I feel dont need to so explicitly show what will be returned since the toString method is quite clear

JodyLorah reviewed (commented):
Generally looks good and neat, but javadocs lacking in some classes,

markmcwong commented on src/main/java/chat/Ui.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Prints lines that divide each response.
+     */
+    public void lines() {
```
Perhaps naming the method as printLine() since it is not operating on a collection of objects (not lines but rather only one line) and use verb at the start also?

markmcwong commented on src/main/java/chat/Parser.java:
```diff
@@ -0,0 +1,38 @@
+     * @throws ChatException If command is not known by Chat the Cat.
+     */
+    public static Command parse(String str) throws ChatException {
+        if (str.equals("list")) {
+            return new ListCommand();
+        } else if (str.startsWith("done")) {
+            return new DoneCommand(str);
+        } else if (str.startsWith("todo") | str.startsWith("deadline") | str.startsWith("event")) {
+            return new AddCommand(str);
+        } else if (str.startsWith("delete")) {
+            return new DeleteCommand(str);
+        } else if (str.equals("bye")) {
+            return new ExitCommand();
+        } else if (str.startsWith("find")) {
+            return new FindCommand(str);
+        } else {
```
Perhaps using a switch statement will be clear and easier to maintain?

clarlzx replied:
Yep I think that will be better, thanks for the suggestion 😄.

markmcwong commented on src/main/java/chat/command/FindCommand.java:
```diff
@@ -0,0 +1,47 @@
+package chat.command;
+
+import chat.*;
```
Perhaps avoid using the wild imports and list imported classes explicitly?

clarlzx replied:
Yes that seems better, thank you for pointing that out.

markmcwong commented on src/main/java/chat/task/Deadline.java:
```diff
@@ -0,0 +1,133 @@
+ */
+public class Deadline extends Task {
+
+    private LocalDateTime end;
```
Perhaps naming "end" as a more descriptive name such endDate ?

markmcwong commented on src/main/java/chat/task/Deadline.java:
```diff
@@ -0,0 +1,133 @@
+            //i.e. deadline return book tmr
+            throw new ChatException("/by missing\n" + formatStr);
+        } else if (!tempStr.contains(" /by ")) {
+            //i.e. deadline return book/by
+            //i.e. deadline return book/by tmr
+            //i.e. deadline return book /bytmr
+            //i.e. deadline return book/bytmr
```
nice description of what the `if` condition will filter out with examples

markmcwong reviewed (commented):
clear and well documented, with examples in comments to make readers easier to understand

clarlzx committed:
A-CheckStyle: Add checkStyle to detect coding style violations

nicholasnge commented on src/main/java/chat/Chat.java:
```diff
@@ -0,0 +1,64 @@
+     * Method that starts and runs Chat the Cat.
+     */
+    public void run() {
+        ui.lines();
```
Could consider calling lines() in the greet() method (and other places where you call it) instead of in the main, where lines are very precious. ie. in the greet method:
`greet() {
ui.lines();
...
ui.lines()
}`

nicholasnge commented on src/main/java/chat/Chat.java:
```diff
@@ -0,0 +1,64 @@
+ * Main class that executes Chat the Cat.
+ */
+
+public class Chat {
```
love the concept of Chat the cat

clarlzx replied:
Hahaha, thanks 😸

nicholasnge commented on src/main/java/chat/Ui.java:
```diff
@@ -0,0 +1,119 @@
+ */
+public class Ui {
+    
+    private static String greetingCat = "(=^. .^=)";
```
incredible stuff

nicholasnge commented on src/main/java/chat/Ui.java:
```diff
@@ -0,0 +1,119 @@
+    public void list(TaskList tasks) {
+        int i = 0;
+        System.out.println(" * list *");
+        while (i < tasks.getTasks().size()) {
```
i think its more standard to use the standard for loop in this scenario, even though they do the same thing... so that the code can be less surprising and more standard

nicholasnge commented on src/main/java/chat/command/Command.java:
```diff
@@ -0,0 +1,70 @@
+     */
+    public Task checkCommandIndex(TaskList tasks, String commandName, String str) throws ChatException {
+        //for commands with format: [command] [index]
+        //returns valid index if index is correct
```
the structure here is abit hard to follow... since _"Please input with this ... "_ occurs quite alot in the exceptions, maybe can refactor that phrase out of all the exceptions which point to formatting error?

clarlzx replied:
Yes I think that this method could be made cleaner, thanks for the suggestion.

nicholasnge commented on src/main/java/chat/task/Deadline.java:
```diff
@@ -0,0 +1,133 @@
+     * Returns a string that shows the details of the task.
+     * <p>[ ] will be displayed if isDone = false.</p>
+     * <p>[X] will be displayed if isDone = true.</p>
+     *
```
seriously very nice docs

nicholasnge commented on src/test/java/chat/task/EventTest.java:
```diff
@@ -0,0 +1,321 @@
+    }
+
+    @Test
+    public void createEvent_noSpaceBetweenEventAndDescription_chatException() {
```
hmm isnt the standard to capitalize the first words, eg. CreateEvent_NoSpace..._ChatException

clarlzx replied:
Maybe you have remembered wrongly because I found this on the coding standard guide?

![image](https://user-images.githubusercontent.com/66005036/106916372-0ee01100-6742-11eb-9648-caaaa51da387.png)

nicholasnge reviewed (commented):
very nice comments and examples in the docs

clarlzx committed:
Level-10: Add GUI to Chat

clarlzx committed:
Merge branch 'branch-Level-10' into master

clarlzx committed:
Edit code based on pull request suggestions

clarlzx committed:
Edit mainClassName in build.gradle

clarlzx committed:
A-Assertions: Add assertions to all classes

clarlzx committed:
A-CodeQuality: Improve code quality of Storage class, Task classes and Command classes

clarlzx committed:
Merge pull request #1 from clarlzx/branch-A-Assertions

Merge branch-A-Assertions to master

clarlzx committed:
Merge branch 'master' into branch-A-CodeQuality and fix merge conflicts

clarlzx committed:
Merge pull request #3 from clarlzx/branch-A-CodeQuality

Merge branch-A-CodeQuality to master

clarlzx committed:
C-BetterSearch: Improve find to search by task type, name, status (done or not done) and dates

clarlzx committed:
A-BetterGui: Create individual ChatDialogBox and UserDialogBox classes

clarlzx committed:
FindCommand: Add more flexibility to find function

Keywords can be arranged in any order.
Keywords are also categorised by keywords such that users will not be
confused when entering the search terms.

clarlzx committed:
Organise imported packages in all classes

clarlzx committed:
Edit Chat's responses

clarlzx committed:
Update documents with README.md and screenshot of user interface

clarlzx committed:
FindCommand: Change delimiter for dates

clarlzx committed:
Change example for find command

clarlzx committed:
Remove all unnecessary files from docs directory

clarlzx committed:
Edit headers in README.md

clarlzx committed:
Edit headers in README.md

clarlzx committed:
Delete README.md

clarlzx committed:
Update base on feedback given during the peer evaluation phase of iP evaluation

clarlzx committed:
Merge branch 'master' of https://github.com/clarlzx/ip into master

clarlzx committed:
Update base on feedback given during the peer evaluation phase of iP evaluation
